src/pubsub.js

```diff
@@ -50,6 +63,76 @@ class BasicPubSub extends Pubsub {
     })
   }
 
+  /**
+   * Dial a received peer.
+   * @override
+   * @param {PeerInfo} peerInfo The peer being dialed
+   * @param {function} callback
+   *
+   * @returns {void}
+   */
+  _dialPeer (peerInfo, callback) {
+    callback = callback || function noop () { }
+    const idB58Str = peerInfo.id.toB58String()
+
+    // If already have a PubSub conn, ignore
+    const peer = this.peers.get(idB58Str)
+    if (peer && peer.isConnected) {
+      return nextTick(() => callback())
+    }
+
+    // If already dialing this peer, ignore
+    if (this._dials.has(idB58Str)) {
+      this.log('already dialing %s, ignoring dial attempt', idB58Str)
+      return nextTick(() => callback())
+    }
+
+    // Verify if is known that the peer does not support Gossipsub
+    let onlySupportsFloodsub = peerInfo.protocols.has(floodsubMulticodec) && !peerInfo.protocols.has(this.multicodec)
+
+    // Define multicodec to use
+    // Should fallback to floodsub if fallback is enabled, protocols were negotiated, and no Gossipsub available
+    let multicodec = this.multicodec
+
+    if (this._options.fallbackToFloodsub && onlySupportsFloodsub) {
+      multicodec = floodsubMulticodec
+    }
+
+    this._dials.add(idB58Str)
+    this.log('dialing %s %s', multicodec, idB58Str)
+
+    this.libp2p.dialProtocol(peerInfo, multicodec, (err, conn) => {
+      this.log('dial to %s complete', idB58Str)
+      this._dials.delete(idB58Str)
+
+      if (err) {
+        // If previously dialed gossipsub and not supported, try floodsub if enabled fallback
+        if (this._options.fallbackToFloodsub &&
+          multicodec === this.multicodec &&
+          err.code === multistreamConstants.errors.MULTICODEC_NOT_SUPPORTED) {
+          this._dials.add(idB58Str)
+          this.log('dialing %s %s', floodsubMulticodec, idB58Str)
+
+          this.libp2p.dialProtocol(peerInfo, floodsubMulticodec, (err, conn) => {
+            this.log('dial to %s complete', idB58Str)
+            this._dials.delete(idB58Str)
+
+            if (err) {
+              this.log.err(err)
+              return callback()
+            }
+            this._onDial(peerInfo, conn, callback)
+          })
+        } else {
+          this.log.err(err)
+          return callback()
```

Comment:
I'm confused about the use of the `callback` in this function. It looks like it's being used as for errors, example:
```
_dialPeer(peerInfo, function() {
    // this is an error
}
```

But typically we wan to pass propagate the error up, and allow the user to handle it. Alternatively we can add two parameters to the the callback `callback(err, value)` and allow the user to handle the error so it looks more like:
```
_dialPeer(peerInfo, function(err) {
    if (err) {
      console.log(err)
      // Do something
    }
}
```
Bare in mind im not familiar with the logger in gossipsub-js but, assuming the log levels are set this could be a silent error.

Reply:
You are having good reasoning since we usually use callbacks to propagate values/errors through async logic. However, in this case, is a little different. I will explain to you the flow:

1) Peer got connected (emit event so that everyone interested in it can handle appropriately)
https://github.com/libp2p/js-libp2p/blob/master/src/index.js#L80-L84

2) Pubsub is waiting for peers to connect, in order to add them to the appropriate pubsub router
https://github.com/libp2p/js-libp2p-pubsub/blob/master/src/index.js#L319

3) Gossipsub (in this case) overrides the `_dialPeer` function, which is called and tries to dial it using its pubsub multicodec
(this function here)

All in all, we don't mind if a peer does not join the gossipsub overlay network. We can have peers that are not running any pubsub router and we just log that we could not add them, there is nothing else the peer can do.

Reply:
Hmm ok, thats interesting. So ultimately, we don't care if they can't connect. Makes sense.